File: backend/src/schemas/thread.py
"""Thread Pydantic schemas for API requests and responses."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, field_validator


class ThreadBase(BaseModel):
    """Base thread schema with common fields."""

    title: str = Field(..., min_length=1, max_length=100, description="Thread title")
    description: str = Field(..., min_length=1, description="Thread description")
    image_url: Optional[str] = Field(None, description="Cover image URL")
    category: str = Field(..., min_length=1, max_length=50, description="Thread category (e.g., gaming, fashion)")
    tags: list[str] = Field(default_factory=list, description="Array of tags")


class ThreadCreate(ThreadBase):
    """Schema for creating a new thread."""

    pass


class ThreadUpdate(BaseModel):
    """Schema for updating an existing thread."""

    title: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, min_length=1)
    image_url: Optional[str] = None
    category: Optional[str] = Field(None, min_length=1, max_length=50)
    tags: Optional[list[str]] = None


class ThreadResponse(ThreadBase):
    """Schema for thread API responses."""

    id: int
    tier: int = Field(..., description="Thread tier level (0-3)")
    contributions: int = Field(..., description="Number of boosts/contributions")
    member_count: int = Field(..., description="Total members in this thread")
    created_by_user_id: int
    created_at: datetime
    updated_at: datetime

    # Add computed fields for frontend
    online_users: int = Field(default=0, description="Currently online users (computed)")

    class Config:
        from_attributes = True  # Pydantic v2 (was orm_mode in v1)


class ThreadListResponse(BaseModel):
    """Schema for listing multiple threads."""

    threads: list[ThreadResponse]
    total: int

File: frontend/src/types/thread.ts
/**
 * Thread types matching backend API response (snake_case from API)
 */

export interface Thread {
  id: number;
  title: string;
  description: string;
  image_url: string | null;
  category: string;
  tags: string[];
  tier: number;
  contributions: number;
  active_contributions: number;
  boost_expires_at: string | null;
  member_count: number;
  created_by_user_id: number;
  created_at: string;
  updated_at: string;
  online_users: number;
}

export interface ThreadListResponse {
  threads: Thread[];
  total: number;
}

/**
 * Thread display type for frontend components (camelCase)
 * Used by UI components and mock data
 */
export interface ThreadDisplay {
  id: number;
  title: string;
  description: string;
  imageUrl: string;
  comments: number;
  views: number;
  upvotes: number;
  currentTokens: number;
  goalTokens: number;
  tags: string[];
  isPinned: boolean;
  isHot: boolean;
  timeAgo: string;
  contributions: number;
  category: string;
  tier?: number;
  onlineUsers?: number;
  totalUsers?: number;
}
